test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/DefaultRoslynCompilationServiceTest.cs

```diff
@@ -0,0 +1,313 @@
+                            Assert.Equal(2, message.StartColumn);
+                            Assert.Equal(11, message.EndLine);
+                            Assert.Equal(3, message.EndColumn);
+                        },
+                        message =>
+                        {
+                            Assert.Equal("message-3", message.Message);
+                            Assert.Equal(viewPath, message.SourceFilePath);
+                            Assert.Equal(31, message.StartLine);
+                            Assert.Equal(6, message.StartColumn);
+                            Assert.Equal(41, message.EndLine);
+                            Assert.Equal(13, message.EndColumn);
+                        });
+                },
+                failure =>
+                {
+                    Assert.Equal(generatedCodeFileName, failure.SourceFilePath);
+                    Assert.Equal("compilation-content", failure.SourceFileContent);
+                    Assert.Collection(failure.Messages,
+                        message =>
+                        {
+                            Assert.Equal("message-2", message.Message);
+                            Assert.Equal(assemblyName, message.SourceFilePath);
+                            Assert.Equal(2, message.StartLine);
+                            Assert.Equal(3, message.StartColumn);
+                            Assert.Equal(4, message.EndLine);
+                            Assert.Equal(5, message.EndColumn);
+                        });
+                });
+        }
+
+        [Fact]
+        public void Compile_RunsCallback()
+        {
+            var content = "public class MyTestType  {}";
+            RoslynCompilationContext usedCompilation = null;
+
+            var compilationService = new DefaultRoslynCompilationService(
+                 GetDependencyContext(),
+                 GetOptions(callback: c => usedCompilation = c),
+                 GetFileProviderAccessor(),
+                 NullLoggerFactory.Instance);
+
+            var relativeFileInfo = new RelativeFileInfo(
+                new TestFileInfo { PhysicalPath = "SomePath" },
+                "some-relative-path");
+
+            // Act
+            var result = compilationService.Compile(relativeFileInfo, content);
+
+            Assert.NotNull(usedCompilation);
+            Assert.Single(usedCompilation.Compilation.SyntaxTrees);
+        }
+
+        private static DiagnosticDescriptor GetDiagnosticDescriptor(string messageFormat)
+        {
+            return new DiagnosticDescriptor(
+                id: "someid",
+                title: "sometitle",
+                messageFormat: messageFormat,
+                category: "some-category",
+                defaultSeverity: DiagnosticSeverity.Error,
+                isEnabledByDefault: true);
+        }
+
+        private static RazorViewEngineOptions GetOptions(Action<RoslynCompilationContext> callback = null)
```

Comment:
Not needed, just inline `new RazorViewEngineOptions()` wherever you have `GetOptions`. The options class handles the `null` callback case.